feature : display input number next to slider

victorcerutti opened the pull request:
See feature request here : https://community.home-assistant.io/t/input-slider-should-show-set-value-and-have-unit/4489

Improve the slider of component `input_number` to have state shown next to it. Displayed state uses the specified `unit_of_measurement` in the configuration file.
<img width="514" alt="capture d ecran 2018-01-16 a 10 00 34" src="https://user-images.githubusercontent.com/3236171/34980389-b0a322f6-faa4-11e7-8f8e-570659132cb4.png">


Also add the state on the humidity slider of the climate component in a design similar to the target temperature value
<img width="390" alt="capture d ecran 2018-01-16 a 10 00 55" src="https://user-images.githubusercontent.com/3236171/34980395-b3bc7db6-faa4-11e7-860c-f71169a806b5.png">

victorcerutti committed:
feature : display input number next to slider

victorcerutti commented:
<img width="504" alt="capture d ecran 2018-01-16 a 10 11 25" src="https://user-images.githubusercontent.com/3236171/34980729-a72d17bc-faa5-11e7-8405-b5c80269cc91.png">
I also checked with long name or long integer.
While it breaks alignement with long numbers, everything looks fine

c727 commented:
please also algin the box text to the right

victorcerutti committed:
Improvements for Box mode of input number

Add unit when available and align the value to the right

victorcerutti commented:
@c727 Just updated the PR

<img width="510" alt="capture d ecran 2018-01-16 a 15 57 26" src="https://user-images.githubusercontent.com/3236171/34995276-43ca4628-fad6-11e7-9e38-f8c5e3d71b5d.png">

balloob commented:
The problem is that this takes up a lot of width. Cards can get as small as ~320px, in which case the box is taking up a lot of space.

andrey-git commented:
Actually as small as 280.5px for a 600px wide screen.

victorcerutti commented:
280px width Before vs After feature addition
<img width="614" alt="capture d ecran 2018-01-17 a 10 48 48" src="https://user-images.githubusercontent.com/3236171/35036286-29e00bfa-fb74-11e7-975d-ec8f830944df.png">

This is indeed an issue a the slider becomes unusable

What about hidding the state part when the width is under a fixed size ?
(I think the slider is still fine with a 320/340px width for the card)
Is this acceptable from a UI perspective ?

andrey-git commented:
Is there a way to do so with css without resorting to js?

c727 commented:
Andrey similar to this: https://github.com/ciotlosm/home-assistant-polymer/blob/hassiodev/panels/hassio/hassio-style.html

Yes, hiding the value on small cards sounds good

andrey-git commented:
@c727 thats pretty dirty as it assumes the 300;600;900 columns which are defined in another place.

victorcerutti commented:
I'm not sure that it can be done using @media queries (hiding an element if the card is less than a certain width). I'm not familiar with Polymer so there may be some kind of solution I'm not aware of.

However this is rather simple in javascript and can be done depending on the slider width.
If it is lower that a certain width (exemple: 100px), then we hide the state.
I think this solution is better because it's linked directly to the element that need space to be usable and will be easier to maintain if the layout changes in the future.

If using javascript is OK, then I will update the PR with a fix for small width cards

victorcerutti committed:
fix for small width cards

hide state if slider width<100px

victorcerutti committed:
add a space between if and (

victorcerutti committed:
Fix value display after a window resize

State value is now hiding or showing correctly after a window resize, using iron-resize component

victorcerutti committed:
fixes eslint validator

victorcerutti commented:
Value box now shown/hidden depending on the slider width is updated after a window resize
I think this feature is complete and ready for review

andrey-git commented on src/state-summary/state-card-input_number.html:
```diff
@@ -41,14 +57,25 @@
 
+  ready() {
+    super.ready();
+    if (this.mode === 'slider') {
```
mode can change later, so it shouldn't be done once

victorcerutti replied:
Fixed in https://github.com/home-assistant/home-assistant-polymer/pull/808/commits/9fcab8681a5de1ef8e3093374e074db04709954a

andrey-git commented on src/state-summary/state-card-input_number.html:
```diff
@@ -86,14 +113,26 @@
 
+  hiddenState() {
+    if (this.$.slider.offsetWidth < 100) {
+      this.$.sliderstate.hidden = true;
```
How does this work? `this.$.sliderstate` is a regular div. What causes `hidden` property to propagate to an attribute?

victorcerutti replied:
This is my first time with polymer but if `this.$.sliderstate` is a div, isn't `this.$.sliderstate.hidden` the html attribute of the div (and not a property) ?

balloob replied:
`hidden` is a global attribute: [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/hidden)

andrey-git commented on src/state-summary/state-card-input_number.html:
```diff
@@ -86,14 +113,26 @@
+  }
+
   stateObjectChanged(newVal) {
     this.min = Number(newVal.attributes.min);
```
Use setProperties to apply all properties together. This will cause less observer calls.
https://www.polymer-project.org/2.0/docs/devguide/model-data#set-property

victorcerutti replied:
Improved in https://github.com/home-assistant/home-assistant-polymer/commit/00b07de52507f0f72a86edd624d7c155c960e584

victorcerutti committed:
Fix a typo in state-card

victorcerutti committed:
Improvement : Apply all properties altogether

As suggered by @andrey-git, I made use of setProperties

victorcerutti committed:
Watch for mode update of input_number

If a mode is updated and the box switches from "box" to "slider", it now also update the visibility of the state value depending on the slider width

victorcerutti committed:
Update width showing state value

balloob commented:
For Chrome we could depend on [`ResizeObserver`](https://developers.google.com/web/updates/2016/10/resizeobserver) available in Chrome 64 ([polyfill](https://github.com/que-etc/resize-observer-polyfill))

victorcerutti committed:
Use ResizeObserver when available

victorcerutti commented:
Last commit added support of ResizeObserver when possible. It works only on Chrome >= 64 for now.
It falls back to a regular `iron-resize` eventListener (which listen to window resize) on other browser.

andrey-git commented on src/state-summary/state-card-input_number.html:
```diff
@@ -86,21 +120,37 @@
+      hiddenbox: (newVal.attributes.mode !== 'box'),
+      hiddenslider: (newVal.attributes.mode !== 'slider'),
+    });
+    if (this.mode !== newVal.mode) {
```
Did you mean `newVal.attributes.mode`?
In any case this will never be true because `this.mode` is already set by setProperty

victorcerutti replied:
Yes of course ! This is a mistake
I think I can always call `this.hiddenState()` after a state change anyway, it's not important to check if mode has changed. If it's not slider it will return false. If it is, it will juste recalculate the need to hide the state value.
If there would be a rare case where value changes too often maybe we could check for a `mode` property update, but I don't think that's a use case for an input number.

victorcerutti replied:
I fixed it in the latest commit: https://github.com/home-assistant/home-assistant-polymer/pull/808/commits/b4eeef2dc33ab4ec89b01b7046e59007a37588de.
Finally decided to go the way of checking for the mode change to be certain to cover every use cases.

victorcerutti committed:
Fix conditionnal mistake

Related @andrey-git review here https://github.com/home-assistant/home-assistant-polymer/pull/808#pullrequestreview-90048982
